kernel/sched.c

```diff
@@ -651,81 +649,70 @@ void z_sched_start(struct k_thread *thread)
 	z_reschedule(&_sched_spinlock, key);
 }
 
-/**
- * @brief Halt a thread
- *
- * If the target thread is running on another CPU, flag it as needing to
- * abort and send an IPI (if supported) to force a schedule point and wait
- * until the target thread is switched out (ISRs will spin to wait and threads
- * will block to wait). If the target thread is not running on another CPU,
- * then it is safe to act immediately.
- *
- * Upon entry to this routine, the scheduler lock is already held. It is
- * released before this routine returns.
- *
- * @param thread Thread to suspend or abort
- * @param key Current key for _sched_spinlock
- * @param terminate True if aborting thread, false if suspending thread
+/* Spins in ISR context, waiting for a thread known to be running on
+ * another CPU to catch the IPI we sent and halt.  Note that we check
+ * for ourselves being halted from inside the loop to prevent
+ * deadlock.
+ */
+static k_spinlock_key_t thread_halt_spin(struct k_thread *thread, k_spinlock_key_t key)
+{
+	while (is_halting(thread)) {
+		if (is_halting(_current)) {
+			halt_thread(_current,
+				    is_aborting(_current) ? _THREAD_DEAD
+				    : _THREAD_SUSPENDED);
+		}
+		k_spin_unlock(&_sched_spinlock, key);
+		arch_spin_relax();
+		key = k_spin_lock(&_sched_spinlock);
+	}
+	z_sched_switch_spin(thread);
```

Comment:
I hate to be a negative Nancy, but I don't think that this fixes the deadlock--it just moves it slightly. One CPU will be stuck in _z_sched_switch_spin()_ while the other will be stuck spinning on __sched_spinlock_.

Reply:
Oh, indeed, this has to happen outside the lock (which makes the code as currently structured sort of messy as I'd have to artificially release it again needlessly, will come up with something).  Looks to me like this is something I broke in that refactoring patch.

But the problem isn't unresolvable: as long as the threads in the cycle all notice the ABORTING flag in finite time, then they all exit in finite time, and the unlocked switch_spin() completes without deadlock.  And because it's being done on a thread owned by our up-stack caller, we're guaranteed not to race on the "how to know when a thread struct is still OK to touch" problem.